### PR TITLE
Rename Rule -> Condition, RuleSet -> Rule

### PR DESCRIPTION
**Pull request by snmgian**

### CODE REVIEW

File: lib/ruy/rule.rb [hunk 1]
```diff
@@ -28,28 +33,74 @@ def set(name, value)
...
+      @fallback = value
     end
 
-    def ==(o)
-      o.kind_of?(Rule) &&
-        conditions == o.conditions
+    # Defines a memoized value.
+    #
+    # The value will be resolved upon the context during evaluation. Let is lazy evaluated, it is
+    # not evaluated until the first condition referencing it is invoked. Once evaluated, it's value
```
Comment (santiagodoldan) [inline lib/ruy/rule.rb]: its value

File: lib/ruy/rule.rb [hunk 2]
```diff
@@ -28,28 +33,74 @@ def set(name, value)
...
+    end
+
+    # TODO document
+    def outcome(value, &block)
+      outcome = Outcome.new(value)
+      outcome.instance_exec(&block) if block_given?
+      @outcomes << outcome
     end
 
+    #def ==(o)
```
Comment (santiagodoldan) [inline lib/ruy/rule.rb]: please, remove commented code

File: lib/ruy/rule.rb [hunk 3]
```diff
@@ -28,28 +33,74 @@ def set(name, value)
...
+      #o.kind_of?(Rule) &&
+        #conditions == o.conditions
+    #end
+
     # Getter method for rules params. It returns all the params without nil objects.
     #
     # @return [Array<Object>]
-    def params
-      @params.compact
+    #def params
```
Comment (santiagodoldan) [inline lib/ruy/rule.rb]: same here

File: lib/ruy/rule.rb [hunk 4]
```diff
@@ -4,11 +4,16 @@ class Rule
...
-    def initialize(*args)
+    def initialize(*params)
       @attrs = {}
       @conditions = []
-      @params = args
+      @fallback = nil
+      @lets = []
+      @outcomes = []
+
+      @params = params
```
Comment (santiagodoldan) [inline lib/ruy/rule.rb]: seems that @params is not being used within this class

File: lib/ruy/rule.rb [hunk 5]
```diff
@@ -28,28 +33,74 @@ def set(name, value)
...
+      end
+
+      if @fallback
+        s << "\n" unless s == ''
+        s << "fallback #{@fallback.inspect}\n"
+      end
+
+      s
     end
+
```
Comment (santiagodoldan) [inline lib/ruy/rule.rb]: remove one newline

File: lib/ruy/rule.rb [hunk 6]
```diff
@@ -28,28 +33,74 @@ def set(name, value)
...
     end
 
-    # Evaluates all conditions.
-    #
-    # @return [true] When all conditions succeeds
-    # @return [false] Otherwise
-    def call(ctx)
-      success = @conditions.take_while do |condition|
-        condition.call(ctx)
+    # @param [Hash] context
```
Comment (santiagodoldan) [inline lib/ruy/rule.rb]: missing description

File: lib/ruy/utils/rules.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
+module Ruy
+  module Utils
+    module Rules
+
+      def self.evaluate_conditions(conditions, ctx)
```
Comment (santiagodoldan) [inline lib/ruy/utils/rules.rb]: add a TODO here

File: lib/ruy/utils/rules.rb [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        outcomes.each do |outcome|
+          result = outcome.call(ctx)
+          unless result.nil?
+            return result
+          end
+        end
+
+        nil
+      end
+
```
Comment (santiagodoldan) [inline lib/ruy/utils/rules.rb]: remove one newline

File: spec/lib/ruy/conditions/condition_spec.rb [hunk 9]
```diff
@@ -0,0 +1,37 @@
+require 'spec_helper'
+
+describe Ruy::Conditions::Condition do
+  subject(:condition) { described_class.new }
+
```
Comment (santiagodoldan) [inline spec/lib/ruy/conditions/condition_spec.rb]: remove one newline

File: spec/lib/ruy/conditions/condition_spec.rb [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+      end
+
+      it 'returns true when conditions meet' do
+        raw_ctx[:flag1] = true
+        raw_ctx[:flag2] = true
+
+        expect(condition.call(ctx)).to be(true)
+      end
+
+      it 'returns true when some condition does not meet' do
```
Comment (santiagodoldan) [inline spec/lib/ruy/conditions/condition_spec.rb]: returns false when at least one condition does not meet ?

File: README.md [hunk 11]
```diff
@@ -183,13 +192,13 @@ You cannot use matchers inside nested blocks in a `tz` block expecting them to w
 A possible workaround for this is to use `tz` blocks inside the nested block in question:
 
 ```ruby
-ruleset = Ruy::RuleSet.new
+rule = Ruy::Rule.new
 any do
```
Comment (santiagodoldan) [inline README.md]: rule.any

File: lib/ruy/rule.rb (hunk 1 above)
Comment (snmgian) [inline lib/ruy/rule.rb]: Done

File: lib/ruy/rule.rb (hunk 2 above)
Comment (snmgian) [inline lib/ruy/rule.rb]: Done

File: lib/ruy/rule.rb (hunk 3 above)
Comment (snmgian) [inline lib/ruy/rule.rb]: Done

File: lib/ruy/rule.rb (hunk 4 above)
Comment (snmgian) [inline lib/ruy/rule.rb]: Done

File: lib/ruy/rule.rb (hunk 5 above)
Comment (snmgian) [inline lib/ruy/rule.rb]: Done

File: lib/ruy/rule.rb (hunk 6 above)
Comment (snmgian) [inline lib/ruy/rule.rb]: Done

File: lib/ruy/utils/rules.rb (hunk 7 above)
Comment (snmgian) [inline lib/ruy/utils/rules.rb]: Documentation added

File: lib/ruy/utils/rules.rb (hunk 8 above)
Comment (snmgian) [inline lib/ruy/utils/rules.rb]: Done

File: spec/lib/ruy/conditions/condition_spec.rb (hunk 9 above)
Comment (snmgian) [inline spec/lib/ruy/conditions/condition_spec.rb]: Done

File: spec/lib/ruy/conditions/condition_spec.rb (hunk 10 above)
Comment (snmgian) [inline spec/lib/ruy/conditions/condition_spec.rb]: Done

File: README.md (hunk 11 above)
Comment (snmgian) [inline README.md]: Done

**Pull request merged**